bin/init_exercise.py

```diff
@@ -0,0 +1,411 @@
+#!/usr/bin/env python
+"""
+Script to initialize an exercise for Exercism's Rust track
+
+Why a Python script in the Rust track repo? Distribution.
+A rust program would either need to be precompiled for various
+platforms, and available for download (in which case it wouldn't
+conveniently work in the repository), or would need to be included
+as a sub-crate and compiled locally. A python script can simply
+be a single file and depend on the user's system Python, if desired.
+
+This module requires Python3.5 or newer
+"""
+from __future__ import print_function
+
+try:
+    import collections
+    import json
+    import os
+    import shlex
+    import subprocess
+    import string
+    import sys
+    import urllib.request
+
+    from contextlib import contextmanager
+    from datetime import datetime
+    from uuid import uuid4
+except ImportError:
+    print("This script requires Python 3.5 or higher", file=sys.stderr)
+    # exiting like this isn't great for library use, but at least it's a quick fail
+    sys.exit(1)
+
+# check version info
+if sys.version_info[0] != 3 or sys.version_info[1] < 5:
+    print("This script requires Python 3.5 or higher", file=sys.stderr)
+    # exiting like this isn't great for library use, but at least it's a quick fail
+    sys.exit(1)
+
+
+def output_of(cmd, check_returncode=True):
+    "Return the stdout of the given command"
+    sp = subprocess.run(shlex.split(cmd),
+                        stdout=subprocess.PIPE,
+                        universal_newlines=True)
+    if check_returncode:
+        sp.check_returncode()
+    return sp.stdout.strip()
+
+
+REPO_ROOT = output_of('git rev-parse --show-toplevel')

```

Comment:
Potential worry if the script is run when CWD is outside of the rust repo (including possibly in any other repo)?

Reply:
This script will create a Rust exercise in the `exercises` directory of whatever repo the CWD is currently in when the script is run. If the CWD is not in a git repo, the script breaks. If the CWD is in a repo without an `exercises` directory, the script breaks. No attempts are made to do anything smart in those cases; it just prints the traceback and exits. 

In practice, I can't see anyone ever actually running this from within a non-Rust repo; I think the worry is more theoretical than practical. The bits which insert the entry into `config.json` are interactive anway, meaning that if someone runs this as a script they'll use the `--dont-update-config` flag, which means they can just `cp` or `mv` or `rm` the generated directory appropriately.

Reply:
One might consider a solution that is not sensitive to where the CWD is: the script determines its own location, and acts on the config.json and exercises directories at the proper relative location to itself.

then we run into problems if the person running the script has moved it before running it.

I will say that I am far more likely to run this with CWD being outside of the repo than I am to move this script.

Reply:
Yeah, I'm not too worried about people moving the script. 

However, something like `os.chdir(os.path.dirname(__file__))` before this line would probably fix both issues. I'll look into it.

Reply:
Just pushed a version which does this; local testing confirms that the script now works correctly when CWD is outside the repo.